### PR TITLE
Add gcc 11 support

### LINKED ISSUE #1378: CPU test cases break with gcc 11.x
Compiling with the recent gcc version breaks our test cases. I assume that this is because of internal libstdc++ changes in how it handles `std::function`. This bug affects the Omp2Threads, Boost.Fiber and `std::thread` back-ends. Example error message:

```
/usr/include/c++/11.1.0/type_traits: In instantiation of 'constexpr bool std::__call_is_nt(std::__invoke_other) [with _Fn = alpaka::AccCpuOmp2Threads<std::integral_constant<long unsigned int, 2>, unsigned int>::AccCpuOmp2Threads<alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int> >(const alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int>&, const size_t&)::<lambda()>&; _Args = {}]':
/usr/include/c++/11.1.0/type_traits:2891:34:   required from 'struct std::__call_is_nothrow<std::__invoke_result<alpaka::AccCpuOmp2Threads<std::integral_constant<long unsigned int, 2>, unsigned int>::AccCpuOmp2Threads<alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int> >(const alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int>&, const size_t&)::<lambda()>&>, alpaka::AccCpuOmp2Threads<std::integral_constant<long unsigned int, 2>, unsigned int>::AccCpuOmp2Threads<alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int> >(const alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int>&, const size_t&)::<lambda()>&>'
/usr/include/c++/11.1.0/type_traits:138:12:   required from 'struct std::__and_<std::__is_nt_invocable_impl<std::__invoke_result<alpaka::AccCpuOmp2Threads<std::integral_constant<long unsigned int, 2>, unsigned int>::AccCpuOmp2Threads<alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int> >(const alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int>&, const size_t&)::<lambda()>&>, bool, void>, std::__call_is_nothrow<std::__invoke_result<alpaka::AccCpuOmp2Threads<std::integral_constant<long unsigned int, 2>, unsigned int>::AccCpuOmp2Threads<alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int> >(const alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int>&, const size_t&)::<lambda()>&>, alpaka::AccCpuOmp2Threads<std::integral_constant<long unsigned int, 2>, unsigned int>::AccCpuOmp2Threads<alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int> >(const alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int>&, const size_t&)::<lambda()>&> >'
/usr/include/c++/11.1.0/type_traits:2986:12:   required from 'struct std::is_nothrow_invocable_r<bool, alpaka::AccCpuOmp2Threads<std::integral_constant<long unsigned int, 2>, unsigned int>::AccCpuOmp2Threads<alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int> >(const alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int>&, const size_t&)::<lambda()>&>'
/usr/include/c++/11.1.0/type_traits:3016:54:   required from 'constexpr const bool std::is_nothrow_invocable_r_v<bool, alpaka::AccCpuOmp2Threads<std::integral_constant<long unsigned int, 2>, unsigned int>::AccCpuOmp2Threads<alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int> >(const alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int>&, const size_t&)::<lambda()>&>'
/usr/include/c++/11.1.0/bits/invoke.h:105:14:   required from 'constexpr std::enable_if_t<is_invocable_r_v<_Res, _Callable, _Args ...>, _Res> std::__invoke_r(_Callable&&, _Args&& ...) [with _Res = bool; _Callable = alpaka::AccCpuOmp2Threads<std::integral_constant<long unsigned int, 2>, unsigned int>::AccCpuOmp2Threads<alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int> >(const alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int>&, const size_t&)::<lambda()>&; _Args = {}; std::enable_if_t<is_invocable_r_v<_Res, _Callable, _Args ...>, _Res> = bool]'
/usr/include/c++/11.1.0/bits/std_function.h:291:30:   [ skipping 5 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
/home/stepha27/workspace/caravan/alpaka/include/alpaka/queue/QueueGenericThreadsNonBlocking.hpp:155:64:   required from 'static void alpaka::traits::Enqueue<alpaka::QueueGenericThreadsNonBlocking<TDev>, TTask>::enqueue(alpaka::QueueGenericThreadsNonBlocking<TDev>&, const TTask&) [with TDev = alpaka::DevCpu; TTask = alpaka::TaskKernelCpuOmp2Threads<std::integral_constant<long unsigned int, 2>, unsigned int, MandelbrotKernel, unsigned int*, const unsigned int&, const unsigned int&, unsigned int, const float&, const float&, const float&, const float&, const unsigned int&>]'
/home/stepha27/workspace/caravan/alpaka/include/alpaka/queue/Traits.hpp:48:62:   required from 'void alpaka::enqueue(TQueue&, TTask&&) [with TQueue = alpaka::QueueGenericThreadsNonBlocking<alpaka::DevCpu>; TTask = const alpaka::TaskKernelCpuOmp2Threads<std::integral_constant<long unsigned int, 2>, unsigned int, MandelbrotKernel, unsigned int*, const unsigned int&, const unsigned int&, unsigned int, const float&, const float&, const float&, const float&, const unsigned int&>&]'
/home/stepha27/workspace/caravan/alpaka/include/alpaka/test/MeasureKernelRunTime.hpp:39:32:   required from 'std::chrono::duration<long int, std::ratio<1, 1000> >::rep alpaka::test::integ::measureTaskRunTimeMs(TQueue&, TTask&&) [with TQueue = alpaka::QueueGenericThreadsNonBlocking<alpaka::DevCpu>; TTask = const alpaka::TaskKernelCpuOmp2Threads<std::integral_constant<long unsigned int, 2>, unsigned int, MandelbrotKernel, unsigned int*, const unsigned int&, const unsigned int&, unsigned int, const float&, const float&, const float&, const float&, const unsigned int&>&; std::chrono::duration<long int, std::ratio<1, 1000> >::rep = long int]'
/home/stepha27/workspace/caravan/alpaka/test/integ/mandelbrot/src/mandelbrot.cpp:335:81:   required from 'void ____C_A_T_C_H____T_E_M_P_L_A_T_E____T_E_S_T____F_U_N_C____0() [with TestType = alpaka::AccCpuOmp2Threads<std::integral_constant<long unsigned int, 2>, unsigned int>]'
/home/stepha27/workspace/caravan/alpaka/test/integ/mandelbrot/src/mandelbrot.cpp:259:1:   required from 'void {anonymous}::ns_____C_A_T_C_H____T_E_M_P_L_A_T_E____T_E_S_T____1::____C_A_T_C_H____T_E_M_P_L_A_T_E____T_E_S_T____1<Types>::reg_tests() [with Types = {alpaka::AccCpuSerial<std::integral_constant<long unsigned int, 2>, unsigned int>, alpaka::AccCpuThreads<std::integral_constant<long unsigned int, 2>, unsigned int>, alpaka::AccCpuFibers<std::integral_constant<long unsigned int, 2>, unsigned int>, alpaka::AccCpuTbbBlocks<std::integral_constant<long unsigned int, 2>, unsigned int>, alpaka::AccCpuOmp2Blocks<std::integral_constant<long unsigned int, 2>, unsigned int>, alpaka::AccCpuOmp2Threads<std::integral_constant<long unsigned int, 2>, unsigned int>}]'
/home/stepha27/workspace/caravan/alpaka/test/integ/mandelbrot/src/mandelbrot.cpp:259:1:   required from here
/usr/include/c++/11.1.0/type_traits:2885:14: error: noexcept-expression evaluates to 'false' because of a call to 'alpaka::AccCpuOmp2Threads<std::integral_constant<long unsigned int, 2>, unsigned int>::AccCpuOmp2Threads<alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int> >(const alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int>&, const size_t&)::<lambda()>' [-Werror=noexcept]
 2885 |       return noexcept(std::declval<_Fn>()(std::declval<_Args>()...));
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/stepha27/workspace/caravan/alpaka/include/alpaka/alpaka.hpp:19,
                 from /home/stepha27/workspace/caravan/alpaka/test/integ/mandelbrot/src/mandelbrot.cpp:10:
/home/stepha27/workspace/caravan/alpaka/include/alpaka/acc/AccCpuOmp2Threads.hpp:109:19: note: but 'alpaka::AccCpuOmp2Threads<std::integral_constant<long unsigned int, 2>, unsigned int>::AccCpuOmp2Threads<alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int> >(const alpaka::WorkDivMembers<std::integral_constant<long unsigned int, 2>, unsigned int>&, const size_t&)::<lambda()>' does not throw; perhaps it should be declared 'noexcept'
  109 |                   []() { return (::omp_get_thread_num() == 0); })
      |            
```

The error message points to this line (and its equivalents in the other back-ends):

https://github.com/alpaka-group/alpaka/blob/e10354162f138b15495c34410de2c69414655d84/include/alpaka/acc/AccCpuFibers.hpp#L107

The error can be fixed by adding `noexcept` to the lambda we see there. I'm not sure if this is enough for an actual fix or if we need to investigate this further.

### PR DESCRIPTION
**Pull request by j-stephan**
Fixes #1378.

### CODE REVIEW

**Comment by j-stephan** [comment]
If the reviewers are okay with this PR it can be merged with admin superpowers. The failing tests are CI glitches.

File: README.md [hunk 1]
```diff
@@ -70,18 +70,18 @@ Supported Compilers
 
 This library uses C++14 (or newer when available).
 
-|Accelerator Back-end|gcc 5.5 <br/> (Linux)|gcc 6.4/7.3 <br/> (Linux)|gcc 8.1 <br/> (Linux)|gcc 9.1 <br/> (Linux)|gcc 10.3 <br/> (Linux)|clang 4 <br/> (Linux)|clang 5/6/7/8 <br/> (Linux)|clang 9 <br/> (Linux)|clang 10 <br/> (Linux)|clang 11 <br/> (Linux)|clang 12 <br/> (Linux)|Apple LLVM 11.2.1-12.2.0 <br/> (macOS)|MSVC 2019 <br/> (Windows)|
+|Accelerator Back-end|gcc 5.5 <br/> (Linux)|gcc 6.4/7.3 <br/> (Linux)|gcc 8.1 <br/> (Linux)|gcc 9.1 <br/> (Linux)|gcc 10.3 <br/> (Linux)|gcc 11.1 (Linux) <br/>|clang 4 <br/> (Linux)|clang 5/6/7/8 <br/> (Linux)|clang 9 <br/> (Linux)|clang 10 <br/> (Linux)|clang 11 <br/> (Linux)|clang 12 <br/> (Linux)|Apple LLVM 11.2.1-12.2.0 <br/> (macOS)|MSVC 2019 <br/> (Windows)|
```
Comment (bernhardmgruber) [inline README.md]:
```suggestion
|Accelerator Back-end|gcc 5.5 <br/> (Linux)|gcc 6.4/7.3 <br/> (Linux)|gcc 8.1 <br/> (Linux)|gcc 9.1 <br/> (Linux)|gcc 10.3 <br/> (Linux)|gcc 11.1 <br/> (Linux)|clang 4 <br/> (Linux)|clang 5/6/7/8 <br/> (Linux)|clang 9 <br/> (Linux)|clang 10 <br/> (Linux)|clang 11 <br/> (Linux)|clang 12 <br/> (Linux)|Apple LLVM 11.2.1-12.2.0 <br/> (macOS)|MSVC 2019 <br/> (Windows)|
```

File: README.md [hunk 2]
```diff
@@ -70,18 +70,18 @@ Supported Compilers
...
 |---|---|---|---|---|---|---|---|---|---|---|---|---|---|
```
Comment (bernhardmgruber) [inline README.md]:
```suggestion
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
```

**Review by bernhardmgruber — CHANGES_REQUESTED** [review]
The new table does not render correctly on GitHub.

File: README.md (hunk 2 above)
Comment (j-stephan) [reply README.md]: Thanks!

**Review by bernhardmgruber — APPROVED** [review]

**Review by psychocoderHPC — APPROVED** [review]

**Comment by psychocoderHPC** [comment]
since it is a `fix` I added the bug lable

**Pull request merged**